accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,156 @@
+        IEnumerator IEnumerable.GetEnumerator() { ... }
+    }
+```
+
+## Activity.Tags
+
+As mentioned earlier, Activity has Tags property which returns the list of the activity tags. Tags are a list of `KeyValuePair<string, string>`. Tags are proven to be not enough for OT scenarios as we need to support more value types (numeric, bool, numeric array, and nested tags). Activity today has the method `public Activity AddTag(string key, string? value)`. The proposal here is to add more `AddTag` overload accepting more types and add a new getter property return the list of all typed tags. 
+
+```C#
+    public partial class Activity : IDisposable
+    {
+        public void AddTag(string key, int value) { }
+        public void AddTag(string key, double value) { }
+        public void AddTag(string key, bool value) { }
+        public void AddTag(string key, int [] value) { }
+        public void AddTag(string key, double [] value) { }
+        public void AddTag(string key, bool [] value) { } // how interesting this one?
+        public void AddTag(string key, string [] value) { }
+        public void AddTag(string key, ActivityTagsCollection value) { }
+
+        // We cannot change Tags property so we are introducing a new property with a new name
+        public ActivityTagsCollection TypedTags { get; }
+    }
+```
+
+## Proposal Changing already exposed APIs in previous preview releases
+
+As we are introducing the ActivityTagsCollection, we can take advantage of that now and modify some of the exposed APIs to use this collection instead. 
+
+### ActivityEvent
+
+```C#
+    // Already exposed APIs
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityTagsCollection? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityTagsCollection? attributes)
+        public ActivityTagsCollection? Attributes { get; }
+    }
+```
+
+### ActivityLink
+
+```C#
+    // Already exposed APIs
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, ActivityTagsCollection? attributes)
+        public ActivityTagsCollection? Attributes { get; }
```

Comment:
I'm not good at naming things and I'm on the fence here (naming is always the most challenging discussion since everyone has opinion).

It seems we prefer to call it `Attributes` here since `ActivityLink` is a brand new thing which mirrors `SpanLink`, however we decided to go with `Activity.TypedTags` instead of `Activity.Attributes` due to the alignment with existing `ActivityTags`.

Looks like we have three options:
1. Things newly added in .NET 5 will have a consistent name `Attributes` which aligns with OpenTelemetry spec. - downside is that Tag is becoming the weird legacy thing (we can probably have a method `SetAttribute(string, string)` which mirrors what `AddTag` is doing).
2. We will use `Tags` and `TypedTags` consistently across the newly added APIs - downside is that it looks different from OpenTelemetry, I'm guessing it is probably fine since Activity -> Span is already different and the name `Tag` is already established in .NET for years (unless OpenTelemetry later will add another concept called `tags`, which I don't think will ever happen 🤣).
3. The current approach proposed by this PR.

I don't have a strong opinion, personally I would prefer option 1.

Reply:
This is actually a good point. I am inclining to the consistency inside .NET APIs. I mean we should use Tags all the way instead of Attributes. I believe this will be less confusing to the APIs users. even for OT developers will be easy to map .NET Tags to OT Attributes and avoid the confusion mapping 2 names instead of one. I'll modify the proposal with that and let's discuss it more as part of the process.